source/common/conn_pool/conn_pool_base.cc

```diff
@@ -379,10 +408,20 @@ void ConnPoolImplBase::drainConnectionsImpl(DrainBehavior drain_behavior) {
     return;
   }
   closeIdleConnectionsForDrainingPool();
+  if (pending_streams_.empty()) {
+    // The remaining early data clients are non-idle.
+    while (!early_data_clients_.empty()) {
+      ASSERT(early_data_clients_.back()->numActiveStreams() > 0u);
```

Comment:
Out of curiosity, why is this guaranteed to be true?

Reply:
closeIdleConnectionsForDrainingPool() should have closed idle early_data_ready_clients according to my newly added change.